### PR TITLE
[real_tensor_prop] Infer Fake kernels during real tensor prop

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #139213
* #139212

This PR changes real_tensor_prop to also infer fake kernels when the
operator doesn't have it.

We infer the fake output to be of the same properties as the real
output, with unbacked symints in the sizes and some stride order.

Test Plan:
- new tests

### CODE REVIEW

File: torch/_subclasses/fake_tensor.py [hunk 1]
```diff
@@ -2589,3 +2599,66 @@ def dump_cache_stats() -> None:
...
+) -> Any:
+    assert mode.shape_env is not None
+
+    # Only support operators that have all Tensor outputs
+    real_flat_out, spec = pytree.tree_flatten(real_out)
+    if not all(isinstance(t, torch.Tensor) for t in real_flat_out):
+        raise RuntimeError(
+            f"propagate_real_tensors: we don't support operators that return "
+            f"non-Tensors. Got {op._name}"
+        )
```
Comment (pianpwk) [inline torch/_subclasses/fake_tensor.py:2615]: could we just return the primitive values in this case? Besides tensors & those, is it just ScriptObjects that we'd have to deal with?

File: torch/_subclasses/fake_tensor.py [hunk 2]
```diff
@@ -2589,3 +2599,66 @@ def dump_cache_stats() -> None:
...
+        expected = 1
+        fake_stride = expected
+        for s, idx in strides:
+            if s != expected:
+                unsupported(
+                    f"a return was not dense in memory (sizes {real_out.shape} strides {real_out.stride()})"
+                )
+            fake_strides[idx] = fake_stride
+            expected = expected * real_out.shape[idx]
+            fake_stride = fake_stride * fake_shape[idx]
```
Comment (pianpwk) [inline torch/_subclasses/fake_tensor.py:2652]: just for my understanding, why do we do need to have the `.sort()` call above, and why can't we just return the real tensor strides as-is, even if they're not dense in memory?

File: torch/_subclasses/fake_tensor.py (hunk 1 above)
Comment (zou3519) [reply torch/_subclasses/fake_tensor.py:2615]: If an operator returns a non-symbolic float or an int, we'd end up baking it into the graph. So the graph isn't useful for anything beyond the current sample inputs.

Also, most operators do not return non-symbolic floats or ints. If the data in the future ends up showing we need to relax this, we can. But I intended this PR to be an initial conservative version that we can expand on.

**Comment by zou3519** [comment]
Btw, tests are failing because there's a bug somewhere that in some of the export configurations, propagate_real_tensors attaches no real_tensors to the FakeTensors of the graph inputs. I'm working on minimizing the example and figuring that out

File: torch/_subclasses/fake_tensor.py (hunk 2 above)
Comment (zou3519) [reply torch/_subclasses/fake_tensor.py:2655]: >  why do we do need to have the .sort() call above

We're using this to check the tensor is dense in memory. If the tensor has sizes (a, b, c, d), then it is contiguous if the strides look like (b * c * d, c * d, d, 1). The tensor is dense in memory if the tensor can be permuted into being contiguous. We sort the strides to check if they follow that pattern

Comment (pianpwk) [reply torch/_subclasses/fake_tensor.py:2666]: this makes sense, thank you!

**Review by pianpwk — APPROVED** [review]

**Pull request closed without merging**